zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanAPI.java

```diff
@@ -779,55 +773,53 @@ private ActiveScan getActiveScan(JSONObject params) throws ApiException {
+                int id = Integer.parseInt(idString.trim());
+                Plugin scanner = getScannerFromId(policy, id);
+                scanner.setEnabled(enabled);
             }
+        } catch (NumberFormatException e) {
+            log.warn("Failed to parse scanner ID: ", e);
+            throw new ApiException(ApiException.Type.ILLEGAL_PARAMETER, e.getMessage(), e);
         }
     }
 
-    private void setEnabledPolicies(ScanPolicy policy, String[] ids) {
-        policy.getPluginFactory().setAllPluginEnabled(false);
-        if (ids.length > 0) {
+    private void setEnabledCategories(ScanPolicy policy, String[] ids)
+            throws ApiException, ConfigurationException {
+        try {
+            policy.getPluginFactory().setAllPluginEnabled(false);
             for (String id : ids) {
-                try {
-                    int policyId = Integer.valueOf(id.trim());
-                    if (hasPolicyWithId(policyId)) {
-                        for (Plugin scanner : policy.getPluginFactory().getAllPlugin()) {
-                            if (scanner.getCategory() == policyId) {
-                                scanner.setEnabled(true);
-                            }
-                        }
-                    }
-                } catch (NumberFormatException e) {
-                    log.warn("Failed to parse policy ID: ", e);
-                }
+                int categoryId = Integer.parseInt(id.trim());
+                updateRulesOfCategoryInPolicy(categoryId, policy, s -> s.setEnabled(true));
             }
+        } catch (NumberFormatException e) {
+            log.warn("Failed to parse category ID: ", e);
+            throw new ApiException(ApiException.Type.ILLEGAL_PARAMETER, e.getMessage(), e);
         }
     }
 
-    private static boolean hasPolicyWithId(int policyId) {
-        return Arrays.asList(Category.getAllNames()).contains(Category.getName(policyId));
+    private static void verifyCategoryId(int categoryId) throws ApiException {
+        if (categoryId < 0) {
+            throw new ApiException(
+                    ApiException.Type.ILLEGAL_PARAMETER, Integer.toString(categoryId));
+        }
+        if (!Arrays.asList(Category.getAllNames()).contains(Category.getName(categoryId))) {
+            throw new ApiException(ApiException.Type.DOES_NOT_EXIST, Integer.toString(categoryId));
+        }
     }
 
-    private int getPolicyIdFromParamId(JSONObject params) throws ApiException {
-        final int id = getParam(params, PARAM_ID, -1);
-        if (id == -1) {
-            throw new ApiException(ApiException.Type.ILLEGAL_PARAMETER, PARAM_ID);
-        }
-        if (!hasPolicyWithId(id)) {
-            throw new ApiException(ApiException.Type.DOES_NOT_EXIST, PARAM_ID);
+    private void updateRulesOfCategoryInPolicy(
+            int categoryId, ScanPolicy policy, Consumer<Plugin> consumer)
+            throws ApiException, ConfigurationException {
+        verifyCategoryId(categoryId);
```

Comment:
Could be removed if called by `setEnabledCategories`, other calls to `updateRulesOfCategoryInPolicy` are calling `verifyCategoryId` beforehand.